cranelift/isle/isle/src/serialize.rs

```diff
@@ -372,10 +390,19 @@ impl<'a> Decomposition<'a> {
         // no unhandled rules left, or because every candidate for control flow
         // for the remaining rules has already been matched by some ancestor in
         // the tree.
-        debug_assert!(self
-            .candidates
-            .iter()
-            .all(|c| c.state == BindingState::Matched));
+        debug_assert!(self.candidates.iter().all(|c| {
+            if let &Candidate {
+                kind: HasControlFlow::Equal(x, y),
+                ..
+            } = c
+            {
+                let x = self.equal.find(x);
+                let y = self.equal.find(y);
+                x.zip(y).filter(|(x, y)| x == y).is_some()
```

Comment:
A helper on `equal` might make this idiom a little clearer: it's basically asserting that `x` and `y` have overlapping equivalence classes (at least one element in common), right? Maybe something like `self.equal.overlaps(x, y)`?

Reply:
Also now that I look at it again: is it meant to be `x.zip(y)`, that is, find if there's an overlap but only if the identical element is at the same position in both sequences, or is it meant instead to be an all-pairs check?

Reply:
I _think_ I've confused you because I used `Option::zip`, not `Iterator::zip`. The `Iterator` for `Option` returns zero or one results for `None` or `Some`, respectively, so the two are equivalent, but it might be easier to think of this as more like a logical-and.

So this says, if `DisjointSets::find` returns `Some(representative)` of a set for both lookups, and the representative is the same for both, then return `true`.

Reply:
Ah! Right, I had somehow thought this was an iteration over the whole equivalence class. That makes much more sense, thanks.